cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -114,19 +114,42 @@ func PerformPostUpgradeTasks(client clientset.Interface, cfg *kubeadmapi.MasterC
 	return errors.NewAggregate(errs)
 }
 
-func removeOldKubeDNSDeploymentIfCoreDNSIsUsed(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error {
-	if features.Enabled(cfg.FeatureGates, features.CoreDNS) {
+// removeOldDNSDeploymentIfNewDNSIsUsed removes previously installed dns deployment.
+// If CoreDNS is installed, kube-dns deployment is removed.
+// If kube-dns is installed, CoreDNS deployment is removed.
+func removeOldDNSDeploymentIfNewDNSIsUsed(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error {
+	preCheckKubeDNSDeployment, err := client.AppsV1().Deployments(metav1.NamespaceSystem).Get(kubeadmconstants.KubeDNS, metav1.GetOptions{})
+	if err != nil {
```

Comment:
404 is an apierror, right? I think you don't want to error out in case one of the deployments does not exist.

Reply:
@kargakis please let me know what would be the correct way to implement this.

Reply:
See how it's used elsewhere, eg. [here](https://github.com/kubernetes/kubernetes/blob/dd9981d038012c120525c9e6df98b3beb3ef19e1/pkg/controller/bootstrap/tokencleaner.go#L126).

I think the minimum fix for https://github.com/kubernetes/kubeadm/issues/641 is to ignore 404s when the kubeDNS deployment is deleted.

I would address @xiangpengzhao comment in a separate PR where you need to make sure this function is working properly in all cases below:
* both deployments exist
* only kubeDNS exists
* only coreDNS exists

I would also like to see a unit test that exercises all cases above.

`EnsureDNSAddon` has created the desired DNS deployment at this point so we don't need to handle edge cases like:
* only kubeDNS exists but coreDNS feature gate is true
* only coreDNS exists but coreDNS feature gate is false

Reply:
Thank you @kargakis. I have reverted my changes to only include the minimum fix and as suggested I shall open up a new PR to implement @xiangpengzhao suggestions once this PR has been merged.